PHPRefactoring/phprefactoring.cpp

```diff
@@ -0,0 +1,306 @@
+    if(method.IsEmpty()) {
+        return;
+    }
+
+    if(method.Contains(" ")) {
+        ::wxMessageBox(_("Methode name may not contain spaces"), "PHPRefactoring", wxICON_ERROR | wxOK | wxCENTER);
+        return;
+    }
+
+    std::string range = std::to_string(startLine) + "-" + std::to_string(endLine);
+    wxString parameters = range + " " + method;
+    RefactorFile("extract-method", parameters, editor);
+}
+
+void PHPRefactoring::OnRenameLocalVariable(wxCommandEvent& e)
+{
+    RenameVariable("rename-local-variable");
+}
+
+void PHPRefactoring::OnRenameClassProperty(wxCommandEvent& e)
+{
+    RenameVariable("rename-property");
+}
+
+void PHPRefactoring::RenameVariable(const wxString& action)
+{
+    IEditor* editor = m_manager->GetActiveEditor();
+    if(!editor) {
+        return;
+    }
+
+    wxString line = std::to_string(editor->GetCurrentLine() + 1);
+    wxString oldName = editor->GetWordAtCaret();
+    if(oldName.IsEmpty()) {
+        return;
+    }
+
+    wxString newName = wxGetTextFromUser("New name for " + oldName);
+    if(newName.IsEmpty()) {
+        return;
+    }
+
+    wxString parameters = line + " " + oldName + " " + newName;
+    RefactorFile(action, parameters, editor);
+}
+
+void PHPRefactoring::OnConvertLocalToInstanceVariable(wxCommandEvent& e)
+{
+    IEditor* editor = m_manager->GetActiveEditor();
+    if(!editor) {
+        return;
+    }
+
+    wxString line = std::to_string(editor->GetCurrentLine() + 1);
+    wxString oldName = editor->GetWordAtCaret();
+    if(oldName.IsEmpty()) {
+        return;
+    }
+
+    wxString parameters = line + " " + oldName;
+    RefactorFile("convert-local-to-instance-variable", parameters, editor);
+}
+
+void PHPRefactoring::OnRenameClassAndNamespaces(wxCommandEvent& e)
+{
+    IEditor* editor = m_manager->GetActiveEditor();
+    if(!editor) {
+        return;
+    }
+
+    wxString dir = editor->GetFileName().GetPath();
+    RunCommand("fix-class-names " + dir);
+}
+
+void PHPRefactoring::OnOptimizeUseStatements(wxCommandEvent& e)
+{
+    IEditor* editor = m_manager->GetActiveEditor();
+    if(!editor) {
+        return;
+    }
+
+    RefactorFile("optimize-use", "", editor);
+}
+
+void PHPRefactoring::RefactorFile(const wxString& action, const wxString& extraParameters, IEditor*& editor)
+{
+    wxString filePath, tmpfile, parameters, output;
+    filePath = editor->GetFileName().GetFullPath();
+    tmpfile << filePath << ".php-refactoring-browser";
+    wxString oldContent = editor->GetEditorText();
+    if(!FileUtils::WriteFileContent(tmpfile, oldContent)) {
+        ::wxMessageBox(_("Can not refactor file:\nFailed to write temporary file"), "PHP Refactoring",
+            wxICON_ERROR | wxOK | wxCENTER);
+        return;
+    }
+
+    // Ensure that the temporary file is deleted once we are done with it
+    FileUtils::Deleter fd(tmpfile);
+
+    ::WrapWithQuotes(tmpfile);
+    parameters = action + " " + tmpfile + " " + extraParameters + " | patch -p1";
```

Comment:
I make use of patch to apply the changes to the temp php file. I'm not sure how well this will work on Windows so probably it would be better to return the output and do the patching in the CL code. I noticed that CL already uses the dtl diff libery that might provide the needed functionality, but I'm unsure how to use it.

Not that for the refactoring that changes Namespaces to be PSR-0 complient (fix-class-names) the patch can contain changes to multiple files.